Homework #1. Task #3. Волынец Арсений Викторович

vol0n opened the pull request.

vol0n committed:
Initial commit

vol0n committed:
Initial commit

vol0n committed:
Initial commit

vol0n committed:
Remove taks_1 and task_2

nbirillo commented on src/main/kotlin/homework_1/task_3.kt:
```diff
@@ -0,0 +1,56 @@
+interface Action {
```
Нет пакета

nbirillo commented on src/main/kotlin/homework_1/task_3.kt:
```diff
@@ -0,0 +1,56 @@
+    init {
+        ls.add(0, elem)
+    }
+    override fun cancelAction(ls: MutableList<Int>) = ls.removeFirst().let {}
```
А зачем `let {}`?

vol0n replied:
remove возвращает элемент и не совпадает тип возвращаемого значения с тем, что в интерфейсе указал

nbirillo replied:
Просто смотрится странно. Возможно здесь было бы лучше использовать нотацию  со скобками:
```
override fun cancelAction(ls: MutableList<Int>) {
    ls.removeFirst()
}
```

nbirillo commented on src/main/kotlin/homework_1/task_3.kt:
```diff
@@ -0,0 +1,56 @@
+
+class AddBackAction(elem: Int, ls: MutableList<Int>) : Action {
+    init { ls.add(ls.size, elem) }
+    override fun cancelAction(ls: MutableList<Int>) = ls.removeLast().let {}
```
А зачем `let {}`?

nbirillo commented on src/main/kotlin/homework_1/task_3.kt:
```diff
@@ -0,0 +1,56 @@
+
+class SwapAction(private val i: Int, private val j: Int, ls: MutableList<Int>) : Action {
+    init {
+        val validRange = 0..ls.lastIndex
```
У списка можно получить список индексов

nbirillo commented on src/main/kotlin/homework_1/task_3.kt:
```diff
@@ -0,0 +1,56 @@
+    override fun cancelAction(ls: MutableList<Int>) = ls.removeLast().let {}
+}
+
+class SwapAction(private val i: Int, private val j: Int, ls: MutableList<Int>) : Action {
```
Это же move, поэтому лучше `move` (нужно именно переместить, а не поменять)

nbirillo commented on src/main/kotlin/homework_1/task_3.kt:
```diff
@@ -0,0 +1,56 @@
+    init {
+        val validRange = 0..ls.lastIndex
+        if (i !in validRange || j !in validRange) {
+            // throw Exception("Invalid range for SwapAction")
```
Ну ты либо принти, либо кидай ошибку) не надо оставлять закомментированым какое-то из действий

nbirillo commented on src/main/kotlin/homework_1/task_3.kt:
```diff
@@ -0,0 +1,56 @@
+}
+
+class AddBackAction(elem: Int, ls: MutableList<Int>) : Action {
+    init { ls.add(ls.size, elem) }
```
Можно просто `ls.add(elem)`, по дефолту в конец добавит

nbirillo commented on src/main/kotlin/homework_1/task_3.kt:
```diff
@@ -0,0 +1,56 @@
+            // throw Exception("Invalid range for SwapAction")
+            println("Incorrect indexes for SwapAction: $i, $j. Available range: $validRange")
+        } else {
+            ls.add(j, ls.removeAt(i))
```
Почему ты добавляешь индекс то?... Нужно ведь элемент с этой позиции переместить. Тот же коммента про  отмену действия

nbirillo commented on src/main/kotlin/homework_1/task_3.kt:
```diff
@@ -0,0 +1,56 @@
+}
+
+class PerformedCommandStorage(private val ls: MutableList<Int>) {
+    private val data = emptyList<Action>().toMutableList()
```
Можно сразу создать: `mutableListOf<Action>()`

nbirillo commented on src/main/kotlin/homework_1/task_3.kt:
```diff
@@ -0,0 +1,56 @@
+class PerformedCommandStorage(private val ls: MutableList<Int>) {
+    private val data = emptyList<Action>().toMutableList()
+    fun storeAction(act: Action) = data.add(act)
+    fun cancelAction() = if (data.isEmpty()) println("No actions in the storage, nothing to cancel")
```
Оберни тело функции в {} 
сложно читать такое

nbirillo commented on src/main/kotlin/homework_1/task_3.kt:
```diff
@@ -0,0 +1,56 @@
+        else data.removeAt(data.lastIndex).cancelAction(ls)
+}
+
+/* test
```
Ну можно раскомментировать))

vol0n replied:
там вроде детект ошибки кидает MagicNumber

nbirillo replied:
а, действительно. ну можно для удобства завести константы с этими значениями тут вверху файла к примеру и использовать их. Потом научитесь писать тесты и таких проблем уже не будет) просто постоянно комментировать\убирать комменты неудобно

vol0n committed:
Apply changes

-Add package name
-Rename StoreAction to MoveAction
-Remove let in cancelAction method
-Remove comments

vol0n committed:
Merge branch 'master' of https://github.com/vol0n/spbu_cotlin_hw into hw1_task3

vol0n committed:
Fix conflicts after pull

nbirillo commented on src/main/kotlin/homework1/task_3.kt:
```diff
@@ -0,0 +1,62 @@
+    init {
+        ls.add(0, elem)
+    }
+    override fun cancelAction(ls: MutableList<Int>) {
+        ls.removeFirst()
+    }
```
Это не ошибка, просто комментарий, на будущее, можно еще вот так было (думала может нагуглишь сам):
`override fun cancelAction(ls: MutableList<Int>): Unit = run { ls.removeFirst() }`

nbirillo reviewed (approved):
Зачтено!
